src/cocotb/share/lib/vhpi/VhpiImpl.cpp

```diff
@@ -866,73 +867,103 @@ GpiObjHdl *VhpiImpl::native_check_create(int32_t index, GpiObjHdl *parent) {
+    // instantiation is represented by the object.
+    std::string root_name = vhpi_get_str(vhpiCaseNameP, root);
+    LOG_DEBUG("VHPI: We have found root='%s'", root_name.c_str());
 
-    if (!dut) {
-        if (NULL == (arch = vhpi_handle(vhpiDesignUnit, root))) {
-            LOG_DEBUG("VHPI: Unable to get vhpiDesignUnit via root");
-            check_vhpi_error();
-            return NULL;
-        }
-
-        if (NULL == (dut = vhpi_handle(vhpiPrimaryUnit, arch))) {
-            LOG_DEBUG("VHPI: Unable to get vhpiPrimaryUnit via arch");
-            check_vhpi_error();
-            return NULL;
-        }
+    if (name && !compare_names(name, root_name)) {
+        LOG_DEBUG(
+            "VHPI: root name '%s' doesn't match requested name '%s'. Trying "
+            "fallbacks",
+            root_name.c_str(), name);
+    } else {
+        return create_gpi_obj_from_handle(root, root_name, root_name);
+    }
 
-        /* If this matches the name then it is what we want, but we
-           use the handle two levels up as the DUT as we do not want an
-           object of type vhpiEntityDeclK as the DUT */
+    // Some simulators do not return the correct entity name for rootInst,
+    // so implement fallbacks.
 
-        found = vhpi_get_str(vhpiCaseNameP, dut);
-        dut = root;
+    // First, try to get the entity (primaryUnit) associated with the rootInst
+    // and its name.
+    vhpiHandleT arch = vhpi_handle(vhpiDesignUnit, root);
+    if (!arch) {
+        LOG_DEBUG(
+            "VHPI: Unable to get vhpiDesignUnit (arch) from root handle. "
+            "Trying handle lookup by name");
+        check_vhpi_error();
+    }
 
-    } else {
-        found = vhpi_get_str(vhpiCaseNameP, dut);
+    vhpiHandleT entity = NULL;
+    if (arch && !(entity = vhpi_handle(vhpiPrimaryUnit, arch))) {
+        LOG_DEBUG(
+            "VHPI: Unable to get vhpiPrimaryUnit (entity) from arch handle. "
+            "Trying handle lookup by name");
+        check_vhpi_error();
     }
 
-    if (!dut) {
-        LOG_ERROR("VHPI: Attempting to get the DUT handle failed");
-        return NULL;
+    if (entity) {
+        root_name = vhpi_get_str(vhpiCaseNameP, entity);
+
+        // If this matches the name then it is what we want,
+        // but we use the root handle two levels up as the DUT
+        // because we do not want an object of type vhpiEntityDeclK as the DUT
+        if (name && !compare_names(name, root_name)) {
+            LOG_DEBUG(
+                "VHPI: Root entity name '%s' doesn't match requested name "
+                "'%s'. Trying handle lookup by name",
+                root_name.c_str(), name);
+        } else {
+            return create_gpi_obj_from_handle(root, root_name, root_name);
+        }
     }
 
-    if (!found) {
-        LOG_ERROR("VHPI: Unable to query name for DUT handle");
+    // Second, we search for the root handle by name
+    if (!name) {
+        // name should never by null here, but fail if it is.
```

Comment:
Should never be null here? Isn't this a fallback path if the above searching fails? Is it not possible that the above could fail and the user didn't pass a name? Maybe I'm missing something.

Reply:
If name is null, we'll return the `vhpiRootInst` handle. If we don't get that handle we return a NULL handle without doing the handle lookup by name. That was the previous behavior as well.

So the only way to get here is if a name is passed in for the check, but fails the initial check and the first fallback fails in one of the three ways it could fail.